Single target detection CW

LOCEANlloydizard opened the pull request:
This PR introduces the public API and output schema for single target detection! Ref #1532 

It focuses on:
- defining the function signature and dispatcher integration
- validating required inputs and parameters
- returning a stable xr.Dataset with a target-based schema
- issuing warnings when optional metadata are missing

This provides a foundation for upcoming algorithm development and allows early review of the API design, data model, and test coverage!

LOCEANlloydizard committed:
Refine single target detection implementation

LOCEANlloydizard committed:
Add matecho single-target detection stub and API scaffold

- Remove preliminary Matecho algorithm logic
- Introduce a stub implementation for single-target detection
- Define and validate the public API (inputs, params, outputs)
- Return a zero-target xr.Dataset with a stable schema
- Add parameter and dataset validation checks
- Prepare ground for future algorithmic implementation

LOCEANlloydizard committed:
edit comments

LOCEANlloydizard committed:
Update detect_matecho.py

LOCEANlloydizard committed:
Update test_mask.py

fix tests to match updated single-target detection code

leewujung commented on echopype/tests/mask/test_mask.py:
```diff
@@ -2006,3 +2010,180 @@ def test_echoview_mincan_no_linking():
+        "Heading",
+        "Dist",
+        "TSfreq_matrix",
+    ]
```
changing these to using snake_case?

leewujung replied:
I'll put my comments for the variables themselves in `detect_matecho.py`

leewujung commented on echopype/tests/mask/test_mask.py:
```diff
@@ -2006,3 +2010,180 @@ def test_echoview_mincan_no_linking():
+REQ = {
+    "channel": "chan1",
+    "pulse_length": 1e-3,             # 1 ms dummy
+    "timeSampleInterval_usec": 40.0,  # dummy
```
Not sure where this would be used, but maybe also change this to using snake_case?

leewujung commented on echopype/mask/single_target_detection/detect_matecho.py:
```diff
@@ -0,0 +1,1005 @@
+            target=target,
+            tsfreq_col=np.arange(tsfreq.shape[1], dtype=int),
+            Ping_number=("target", _arr("Ping_number", int)),
+            Time=("target", _arr("Time", np.dtype("datetime64[ns]"))),
```
I don't see any variables above aligned with `Ping_number` or `Time`, so think these are redundant and can be removed?

Also, ping number and time are themselves the same thing, unless the timestamps are not derived from ping time (though I think for ping-by-ping methods like this it _is_ derived from ping time?)

leewujung reviewed (commented):
Hey @LOCEANlloydizard : Thanks for putting together this PR for discussion! I have a few overarching suggestion:
- use xarray data container (dataarray or dataset) instead of dict when possible -- for example at the output of `_compute_core_matrices`, instead of a Dict, a xr.Dataset would work well. BUT see my comment below about that function [1].
- add more inline code comments to enhance code readability
- using xarray operation would help with code readability too, compared to when doing vanilla numpy operations, since it's xarray is label-aware and it's easier to follow the operations.

-------------------------
I have some thoughts between `_cond_8_build_outputs` and `detect_matecho` that would affect the implementation:

If I read the code right, what you have in the double loop in `_cond_8_build_outputs` is to go through all pings (`ip`) that contain detected single targets, and then all ranges (`pidx`) of the detected targets within that ping. The `out` assembly and refinement does 3 things:
1. "flatten" the ping-range structure to just a 1-D array indexed by `target`
2. refine range in meters, though it's still index-based and not on a continuous scale
3. compute some derived variables, such as `positionBeamCoord`

So, at the core the single target detection output has 1 dimension `target` (which is also a coordinate), and 2 variables: the index number along `ping_time`, and the index number along `range_sample` (or `range_meter`, depending on what the detection is operating on [2]). All other information can be extracted or computed from the original dataarray the detection runs on using these 2 index numbers for each target. `isel` or `sel` allow passing multiple indices at once, so seems more efficient than looping (i.e. vectorizing the code).

If the above is true, how about breaking things up into 2 steps? The first function focuses on getting this "core" single target detection output that's flattened (outputA), and the second function focuses on extracting and computing the required values in the full output (outputB)? Then users can decide if they want to store outputA and stop there, or proceed to compute outputB. 

The reasons why I think it may be useful to separate outputA and outputB are:
- it gives a good progression in what the code is trying to do in a workflow
- derived values in outputB likely change more dramatically if calibration or something else change in the source dataset, but outputA is less variable
- I am not exactly sure about this one: this may provide us a way for scaling the single target detection operation in the future, along with the idea that storing an intermediate data product to disk and then read it back for lazy computation is more scalable than trying to do everything in memory at once.

I am also fine with keeping outputA just as an intermediate produce _within_ `detect_matecho` and not exposing it to the user, though thinking it's good to have a structure that's easy to follow.

----------------------------
[1] `_compute_core_matrices` computes quantities that largely overlap with current echopype `compute_Sv` and `compute_TS` functions, and I think it's better to use those directly (and fix any errors there if any!)

[2] I didn't read in detail to see if the current code includes the range decimation you mentioned earlier, so this comment may have no bearing: as part of `compute_Sv/TS` there's a function that computes sample range and store it into the `range_meter` variable in the Sv dataset. Could we compute the refined range from there? This is so that we don't have redundancy in code that does the same thing scattered in different parts of the codebase, which would make maintenance challenging.

LOCEANlloydizard committed:
Merge branch 'main' of https://github.com/OSOceanAcoustics/echopype into single_target_detection

LOCEANlloydizard committed:
Reshape Matecho detector, add EV Method 2, propagate tau_effective

- Refactor / reshape Matecho CW detector interface and outputs
- Propagate tau_effective from compute_Sv (complex encoding path) (Power encoding support still pending)
- Standardise detector return structure (ping_time, range_sample, frequency_nominal
- Additional diagnostic variables (TS, etc.) temporarily retained
- Add minimal Echoview split-beam Method 2 implementation
- Update and refactor associated tests

LOCEANlloydizard committed:
Merge branch 'main' into single_target_detection

LOCEANlloydizard committed:
Merge remote-tracking branch 'upstream/main' into single_target_detection

LOCEANlloydizard committed:
Update branch

Update branch with upstream/main

LOCEANlloydizard committed:
Merge branch 'main' of https://github.com/OSOceanAcoustics/echopype into single_target_detection

LOCEANlloydizard committed:
Update api.py

LOCEANlloydizard committed:
Add prototype API for FM EK80

- Add public `compute_Sv_f()` and `compute_TS_f()` for the API
- Add `_cal_complex_samples_f()` prototype
- Expose `frequency_resolution` and `range_step` parameters for future spectral calibration
- Return prototype `Sv_f` xarray output structure:
  `(channel, ping_time, svf_range, frequency)`

LOCEANlloydizard committed:
Fix calibration dispatcher for unsupported FM methods

Use lazy method getattr in calibration dispatcher to avoid access to unsupported compute_Sv_f / compute_TS_f methods on non-FM calibrators

LOCEANlloydizard committed:
Merge remote-tracking branch 'upstream/main' into broadband_processing

LOCEANlloydizard committed:
Update detect_echoview_split_method2.py

LOCEANlloydizard committed:
major commit for FM TS

- Add compute_TS_spectrum() for EK80 FM complex data
- Add broadband TS spectrum processing pipeline and calibration helpers
- Add CRIMAC reference dataset and validation tests for TS(f)
- Add frequency-dependent absorption validation against CRIMAC
- Add broadband Sp and Sv(f) integration tests
- Add pooch support for ts_spectrum_example_data bundle
- Refactor TS/Sp calibration workflow and API exports
- Add cached test-data handling to avoid unnecessary downloads

LOCEANlloydizard committed:
add tqdm

adding tqdm, most likely temporarily

LOCEANlloydizard committed:
Update conftest.py

LOCEANlloydizard committed:
change org name

LOCEANlloydizard committed:
Update conftest.py

try to fix error occuring only in ubuntu job

LOCEANlloydizard committed:
Fix formatting in test_calibrate_ek80.py

LOCEANlloydizard committed:
removing _get_splitbeam_angles and _get_transducer_halves

LOCEANlloydizard committed:
Merge remote-tracking branch 'upstream/main' into broadband_processing

LOCEANlloydizard committed:
remove Sv(f) API and refactor TS(f) methds

LOCEANlloydizard committed:
removing Sv_f tests

LOCEANlloydizard committed:
append cal_ds_iteration

LOCEANlloydizard committed:
Update api.py

LOCEANlloydizard committed:
Update test_calibrate_ek80_broadband_crimac.py

change the test to a NotImplementedError test

LOCEANlloydizard committed:
Merge branch 'main' into broadband_processing

LOCEANlloydizard committed:
restore conftest.py flattening, after fixing github asset zip file

LOCEANlloydizard committed:
Update conftest.py

LOCEANlloydizard committed:
add beam correction to band-averaged TS

LOCEANlloydizard committed:
Update calibrate_ek.py

LOCEANlloydizard committed:
Remove tqdm and deprecate BB waveform_mode alias

- Remove tqdm import and use
- Deprecate BB waveform_mode alias (keep backward compatibility)
- Change test_fm_equals_bb test to mark "BB" it as deprecated (others would need it too)

LOCEANlloydizard committed:
add test for different windows from scipy

LOCEANlloydizard committed:
add changes and test for split front

LOCEANlloydizard committed:
Merge branch 'broadband_processing' into single_target_detection-merge-broadband

LOCEANlloydizard committed:
update function names

LOCEANlloydizard committed:
Merge branch 'main' into single_target_detection

LOCEANlloydizard committed:
Rename single-target detection methods and update tests

LOCEANlloydizard committed:
update of the api

- change of functions names
- adding FM possible detecton method

LOCEANlloydizard committed:
Merge remote-tracking branch 'upstream/main' into single_target_detection

LOCEANlloydizard committed:
refactor single_target_detection to reflect adding compute_sp

LOCEANlloydizard committed:
Merge branch 'main' into single_target_detection

LOCEANlloydizard committed:
remove "from_Sv" for this PR

LOCEANlloydizard committed:
Adopt SONAR-netCDF-style target variable names
